### ISSUE TITLE
Port twisted.names.hosts to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6092]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6092|
|Type|enhancement|
|Created|2012-10-10 13:32:30Z|
|Branch|https://github.com/twisted/twisted/tree/hosts-py3-6092|

This is (sneakily) needed by `twisted.names.client` (`createResolver` imports it).

<details><summary>Searchable metadata</summary>

```
trac-id__6092 6092
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_hosts_py3_6092 branches/hosts-py3-6092
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1349875950000000 1349875950000000
changetime__1349890927000000 1349890927000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [36050]) Branching to 'hosts-py3-6092'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @itamarst|
|-|-|

That was relatively straightforward.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/hosts-py3-6092).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @exarkun|
|-|-|

Looks good. Some issues:

  1. Missing `__future__`.
  1. You're passing unicode to `FilePath` on Python 3, which isn't really supported in theory, but works in practice. Maybe that's OK though, and it does make things easier for users of the library. Perhaps we should have a ticket for "FilePath can open unicode paths even if it can't yet do child() etc."? Or we can change this to byte paths.

Fix (or not, if you're ok with unicode paths), then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `assigned`|
|-|-|

Thanks!

> Missing `__future__`.

Fixed in r36056.

> You're passing unicode to FilePath on Python 3, which isn't really supported in theory, but works in practice. 

It's a short branch, I know, but it would make the review comment a lot easier to understand if you identified what part of the code you're talking about.  Initially I thought I had forgotten to do some encoding in the new test helper, `GoodTempPathMixin`, then I saw that I hadn't forgotten to do it so I thought you had misread the diff.  Then I thought maybe there was some code somewhere else I'd forgotten about so I grepped the diff.  I didn't see any use of `FilePath` with unicode, but I noticed `hosts.py` importing `FilePath`.  Then I re-examined `hosts.py` for `FilePath` uses and noticed there are some places where unicode really _is_ passed to it - whenever the default value for `hosts.Resolver.__init__`'s `file` parameter is used.  Perhaps that's what this comment is about?  I'll fix that and add a test, too - which will fail, because using unicode with `FilePath` _doesn't_ currently work. ;)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

`Resolver.__init__`, yes but not only. I was less specific (and thought it worked) because I thought you were also passing in unicode in the tests. So I also misread the diff in addition to catching the real problem.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [36059]) Merge hosts-py3-6092

Author: exarkun
Reviewer: itamarst
Fixes: [#6092](https://github.com/twisted/twisted/issues/6092)

Port `twisted.names.hosts` to Python 3.
